### ISSUE TITLE
Key-spacing alignment problem.

### ISSUE THREAD

**Issue by firatkucuk** [eslint/eslint#7661]
I really like intelliJ Idea style object alignment. But I didn't create a linting rule for desired alignment style. For instance the following snippet is properly formatted in My IDE.

```javascript
let items = {
  item115: {
    id      : 100,
    name    : 'item1',
    itemList: [1, 2, 3]
  },
  item2  : {
    id      : 101,
    name    : 'item2',
    itemList: [1, 2, 3]
  }
};
```

My key-spacing rule is:

```json
{
  "key-spacing"    : [
    "error",
    {
      "singleLine": {
        "beforeColon": false,
        "afterColon" : true
      },
      "multiLine" : {
        "beforeColon": true,
        "afterColon" : true,
        "align"      : "colon"
      }
    }
  ]
}
```

IMHO multiLine beforeColon option should be three stated. true, false and off. Is there a way to properly linting above code snippet?

**Comment by eslintbot** [comment]
Hi @firatkucuk, thanks for the issue. It looks like there's not enough information for us to know how to help you.

If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

Requesting a rule change? Please see [Proposing a Rule Change](http://eslint.org/docs/developer-guide/contributing/rule-changes) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by btmills** [comment]
This was last discussed back in #2668, and we had an [accepted proposal](https://github.com/eslint/eslint/issues/2668#issuecomment-140726631), but nobody was interested in implementing it. @firatkucuk, would you be willing to send a PR if we were to decide to do this again?

**Comment by firatkucuk** [comment]
Ok I 'll try to send a PR.

**Comment by btmills** [comment]
@firatkucuk thanks, let's see if we can get support from the team to accept this so you can start working.

---

To the team: We had an accepted proposal in #2668 that would support IntelliJ-style alignment in key-spacing, but the issue was closed when nobody was interested in working on it. @firatkucuk has volunteered to work on a PR, so I'll copy the proposal here and see what you all think.

A new option `"grouping"` which has two possible values:

- `"multi-line"` (default): Values spanning multiple lines end their group
- `"empty-line"`: Blank lines (no comments, no code, just whitespace) create new groups

I do think we should support IntelliJ-style alignment, so I'll champion and throw in a :+1: for this.

**Comment by platinumazure** [comment]
@btmills That's cheating, you can't champion and :+1:. :smile: But I will :+1: this. Need two more!

**Comment by btmills** [comment]
> That's cheating, you can't champion and 👍.

@platinumazure finally! I've never been clear whether it was champion plus three more or the champion counted as one of the three. Thanks for setting me straight 😆  I've edited the comment to avoid confusion

**Comment by firatkucuk** [comment]
I really want to contribute but I am not familiar with the eslint code. Firstly I must find the responsible code and how to run unit test about it. :)

**Comment by btmills** [comment]
@firatkucuk if this is accepted, it'll be my job to point you in the right direction :smile: The [`createGroups` function](https://github.com/eslint/eslint/blob/b30116c36069368a79446adf55367f3ce4a7d271/lib/rules/key-spacing.js#L512) and [`continuesPropertyGroup` predicate](https://github.com/eslint/eslint/blob/b30116c36069368a79446adf55367f3ce4a7d271/lib/rules/key-spacing.js#L36) are the parts that handle the grouping, and all properties in a group have the same alignment. Right now, groups are on [consecutive lines](https://github.com/eslint/eslint/blob/b30116c36069368a79446adf55367f3ce4a7d271/lib/rules/key-spacing.js#L41) or lines separated only by comments. If the team approves it, this change would tweak the grouping logic based on the option.

**Comment by platinumazure** [comment]
@btmills Assigned the issue to you since you're champion. Apologies for not doing this sooner.

**Comment by firatkucuk** [comment]
@btmills @platinumazure I'll do my best.

**Comment by btmills** [comment]
Writing up a more complete explanation of this, as promised a while back:

Currently, when `key-spacing` is configured to align properties vertically, it doesn't necessarily apply the same alignment to every property in the object literal. If properties are on consecutive lines, they will all have the same alignment level. However, if it sees properties separated by whitespace or properties (e.g. function expressions) that span multiple lines, it treats them as different groups.

```js
var obj = {
    a:   1,
	bcd: 3,
	ef:  2
};
```

```js
module.exports = {
    publicProperty: 42,
    publicMethod:   function(foo) {
	    // ...
    },
	yetAnotherPublicProperty: 'foo',
	anotherPublicMethod:      function(bar, baz) {
	    // ...
	}
};
```

In the latter case, the function expressions break up the vertical alignment, so `key-spacing` currently considers them to be separate groups, each with their own alignment levels.

IntelliJ, however, considers those to all have the same alignment level. It breaks groups on empty lines, rather than multi-line values. IntelliJ and `key-spacing` are in conflict here, and there's currently no way to configure `key-spacing` to support IntelliJ's style. The closest way would involve inserting an empty line immediately above `yetAnotherPublicProperty`, which would cause IntelliJ to see two groups just like `key-spacing`, but based on the presence of the empty line, rather than the multi-line bodies.*

**The proposal here is to add a `grouping` option.** The default `"multi-line"` setting would maintain the current behavior. Setting `"grouping": "empty-line"` would instead break groups on empty lines like IntelliJ does.

*I don't use IntelliJ, that's just paraphrased from #2674, so correct me if I'm wrong there.

**Comment by mastilver** [comment]
Isn't that the same as using:

```
'key-spacing': [
      'error',
      {
          align:       'value'
      }
]
```

**Comment by not-an-aardvark** [comment]
@btmills Are you still championing this?

**Comment by btmills** [comment]
Sorry for not keeping track of this. @eslint/eslint-team can we get two more :+1:s? We want to add a `"grouping": "multi-line"/"empty-line"` option to `key-spacing` so we can support IntelliJ-style alignment. See https://github.com/eslint/eslint/issues/7661#issuecomment-267797311 for the full proposal.

**Comment by kaicataldo** [comment]
This proposal is now accepted.

**Comment by nzakas** [comment]
Unfortunately, it looks like there wasn't enough interest from the team or community to implement this change. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to be implemented after 90 days tend to never be implemented, and as such, we [close those issues](https://eslint.org/docs/maintainer-guide/issues#when-to-close-an-issue). This doesn't mean the idea isn't interesting or useful, just that it's not something the team can commit to.
